### PR TITLE
Implement `SquashDictInnerCheckAccessIndex` hint

### LINKED ISSUE #298: SquashDictInnerCheckAccessIndex
Updates the access index during the dictionary squashing process.

Reference: https://github.com/starkware-libs/cairo-lang/blob/efa9648f57568aad8f8a13fbf027d2de7c63c2c0/src/starkware/cairo/common/squash_dict.cairo#L171

Depends on #245

### PR DESCRIPTION
**Pull request by TAdev0**
Resolves #298 

This PR implements `SquashDictInnerCheckAccessIndex` hint, which updates the access index during the dictionary squashing process

### CODE REVIEW

**Comment by TAdev0** [comment]
@rodrigo-pino  ready for review

File: pkg/hintrunner/zero/zerohint_dictionaries.go [hunk 1]
```diff
@@ -89,6 +89,78 @@ func createSquashDictInnerAssertLenKeysHinter() (hinter.Hinter, error) {
 	return newSquashDictInnerAssertLenKeysHint(), nil
 }
 
+// SquashDictInnerCheckAccessIndex hint updates the access index
+// during the dictionary squashing process
+//
+// `newSquashDictInnerCheckAccessIndexHint` takes 1 operander as argument
+//   - `loopTemps` variable is a struct containing a `index_delta_minus1` field
```
Comment (har777) [inline pkg/hintrunner/zero/zerohint_dictionaries.go:96]: can you also say `as the first field`
or maybe address it as a comment below in the code somewhere so the reason for writing to offset `0` of struct is clear

Comment (TAdev0) [reply pkg/hintrunner/zero/zerohint_dictionaries.go]: @har777  just addressed your comment, tell me if it is better like this

File: pkg/hintrunner/zero/zerohint_dictionaries.go [hunk 2]
```diff
@@ -154,6 +154,81 @@ func createSquashDictInnerAssertLenKeysHinter() (hinter.Hinter, error) {
...
+			}
+
+			currentAccessIndex_, err := ctx.ScopeManager.GetVariableValue("current_access_index")
+			if err != nil {
+				return err
+			}
+
+			currentAccessIndex, ok := currentAccessIndex_.(fp.Element)
+			if !ok {
+				return fmt.Errorf("casting currentAccessIndices_ into an array of felts failed")
```
Comment (har777) [inline pkg/hintrunner/zero/zerohint_dictionaries.go:197]: you mean into a felt not an array of felts

File: pkg/hintrunner/zero/zerohint_dictionaries.go [hunk 3]
```diff
@@ -154,6 +154,81 @@ func createSquashDictInnerAssertLenKeysHinter() (hinter.Hinter, error) {
...
+
+			var result fp.Element
+			result.Sub(&newAccessIndex, &currentAccessIndex)
+			result.Sub(&result, &utils.FeltOne)
+
+			resultMem := memory.MemoryValueFromFieldElement(&result)
+
+			// We use 0 as offset for `WriteToNthStructField` function as we write
+			// to the first field of the `loop_temps` struct
+			return vm.Memory.WriteToNthStructField(loopTempsAddr, resultMem, int16(0))
```
Comment (har777) [inline pkg/hintrunner/zero/zerohint_dictionaries.go:218]: If the offset is `0` you can just use `WriteToAddress(loopTempsAddr, resultMem)` also

File: pkg/hintrunner/zero/zerohint_dictionaries.go [hunk 4]
```diff
@@ -154,6 +154,81 @@ func createSquashDictInnerAssertLenKeysHinter() (hinter.Hinter, error) {
...
+
+			currentAccessIndices, ok := currentAccessIndices_.([]fp.Element)
+			if !ok {
+				return fmt.Errorf("casting currentAccessIndices_ into an array of felts failed")
+			}
+
+			newAccessIndex, err := utils.Pop(&currentAccessIndices)
+			if err != nil {
+				return err
+			}
```
Comment (har777) [inline pkg/hintrunner/zero/zerohint_dictionaries.go:188]: I believe you need to update `current_access_indices` in the scope with the new popped array?

Comment (har777) [reply pkg/hintrunner/zero/zerohint_dictionaries.go:188]: As `.pop()` updates the `current_access_indices` object.

File: pkg/hintrunner/zero/zerohint_dictionaries_test.go [hunk 5]
```diff
@@ -96,6 +96,49 @@ func TestZeroHintDictionaries(t *testing.T) {
...
+					if err != nil {
+						t.Fatal(err)
+					}
+				},
+				makeHinter: func(ctx *hintTestContext) hinter.Hinter {
+					return newSquashDictInnerCheckAccessIndexHint(ctx.operanders["loop_temps.index_delta_minus1"])
+				},
+				check: func(t *testing.T, ctx *hintTestContext) {
+					varValueEquals("loop_temps.index_delta_minus1", feltUint64(0))(t, ctx)
+					varValueInScopeEquals("current_access_index", *feltUint64(2))(t, ctx)
```
Comment (har777) [inline pkg/hintrunner/zero/zerohint_dictionaries_test.go:118]: can you also add a check for `current_access_indices` in scope after the pop

**Review by har777 — CHANGES_REQUESTED** [review]

File: pkg/hintrunner/zero/zerohint_dictionaries.go (hunk 3 above)
Comment (TAdev0) [reply pkg/hintrunner/zero/zerohint_dictionaries.go:218]: true. It seems more straightforward, will do that. thanks!

File: pkg/hintrunner/zero/zerohint_dictionaries.go (hunk 2 above)
Comment (TAdev0) [reply pkg/hintrunner/zero/zerohint_dictionaries.go:197]: true thanks !

File: pkg/hintrunner/zero/zerohint_dictionaries_test.go (hunk 5 above)
Comment (TAdev0) [reply pkg/hintrunner/zero/zerohint_dictionaries_test.go:118]: yes right!

File: pkg/hintrunner/zero/zerohint_dictionaries.go (hunk 4 above)
Comment (TAdev0) [reply pkg/hintrunner/zero/zerohint_dictionaries.go:188]: oh damn i forgot that i thought i dit it , and indeed its missing. thanks!

**Comment by TAdev0** [comment]
@har777  just addressed all your comments! thanks for the feedback. Indeed the hint was wrongly implemented, it didnt update the array of felts in scope after popping an element.
I added a test for `current_access_indices` array in scope, now it should be good :)

**Review by har777 — APPROVED** [review]

File: pkg/hintrunner/zero/zerohint_dictionaries.go [hunk 6]
```diff
@@ -396,6 +396,85 @@ func createSquashDictInnerAssertLenKeysHinter() (hinter.Hinter, error) {
...
+// SquashDictInnerCheckAccessIndex hint updates the access index
+// during the dictionary squashing process
+//
+// `newSquashDictInnerCheckAccessIndexHint` takes 1 operander as argument
+//   - `loopTemps` variable is a struct containing `index_delta_minus1` as
+//     a first field
+//
+// `newSquashDictInnerCheckAccessIndexHint` writes to the first field `index_delta_minus1`
+// of the `loop_temps` struct the result of `new_access_index - current_access_index - 1`
+// and assigns `new_access_index` to `current_access_index` in the scope
```
Comment (MaksymMalicki) [inline pkg/hintrunner/zero/zerohint_dictionaries.go:408]: could you add to this annotation an information about how does the loop_temps struct looks like?

File: pkg/hintrunner/zero/zerohint_dictionaries.go [hunk 7]
```diff
@@ -396,6 +396,85 @@ func createSquashDictInnerAssertLenKeysHinter() (hinter.Hinter, error) {
...
+			//> current_access_index = new_access_index
+
+			currentAccessIndices_, err := ctx.ScopeManager.GetVariableValue("current_access_indices")
+			if err != nil {
+				return err
+			}
+
+			currentAccessIndices, ok := currentAccessIndices_.([]fp.Element)
+			if !ok {
+				return fmt.Errorf("casting currentAccessIndices_ into a felt failed")
```
Comment (MaksymMalicki) [inline pkg/hintrunner/zero/zerohint_dictionaries.go:424]: felt array 😛

File: pkg/hintrunner/zero/zerohint_dictionaries_test.go [hunk 8]
```diff
@@ -245,6 +245,49 @@ func TestZeroHintDictionaries(t *testing.T) {
...
+					}
+				},
+				makeHinter: func(ctx *hintTestContext) hinter.Hinter {
+					return newSquashDictInnerCheckAccessIndexHint(ctx.operanders["loop_temps.index_delta_minus1"])
+				},
+				check: func(t *testing.T, ctx *hintTestContext) {
+					varValueEquals("loop_temps.index_delta_minus1", feltUint64(23))(t, ctx)
+					allVarValueInScopeEquals(map[string]any{"current_access_index": *feltUint64(43), "current_access_indices": []fp.Element{*feltUint64(97), *feltUint64(76), *feltUint64(54), *feltUint64(51), *feltUint64(44)}})(t, ctx)
+				}},
+		},
```
Comment (MaksymMalicki) [inline pkg/hintrunner/zero/zerohint_dictionaries_test.go:290]: is it a good ida to test a case with empty `current_access_indices` ?

File: pkg/hintrunner/zero/zerohint_dictionaries.go (hunk 6 above)
Comment (TAdev0) [reply pkg/hintrunner/zero/zerohint_dictionaries.go:408]: sure

File: pkg/hintrunner/zero/zerohint_dictionaries.go (hunk 7 above)
Comment (TAdev0) [reply pkg/hintrunner/zero/zerohint_dictionaries.go:424]: i inversed with another one yeah thanks!

File: pkg/hintrunner/zero/zerohint_dictionaries_test.go (hunk 8 above)
Comment (TAdev0) [reply pkg/hintrunner/zero/zerohint_dictionaries_test.go:290]: just added one more test for that :)

**Comment by TAdev0** [comment]
@MaksymMalicki addressed your comments

**Review by MaksymMalicki — APPROVED** [review]

**Pull request merged**